File: src/lib/auctionRules.ts
import type { Player, TeamManagerEntry } from '../types'

export function getMinAcceptableBid(
  player: Pick<Player, 'currentBid' | 'basePrice'>,
  bidIncrement: number,
): number {
  return player.currentBid > 0 ? player.currentBid + bidIncrement : player.basePrice
}

export function assertValidBid(
  player: Pick<Player, 'currentBid' | 'basePrice' | 'status'>,
  manager: Pick<TeamManagerEntry, 'name' | 'maxPlayers' | 'remainingTokens'>,
  amount: number,
  bidIncrement: number,
  soldCountForManager: number,
  // Number of players this bid would win at once — >1 when the current
  // player is part of a combo lot (see lib/auctions.ts), since the whole
  // group joins the squad together on sale.
  groupSize = 1,
): void {
  if (player.status !== 'open' && player.status !== 'active') {
    throw new Error('Player is not open for bidding')
  }
  const minAcceptable = getMinAcceptableBid(player, bidIncrement)
  if (amount < minAcceptable) {
    throw new Error(`Bid must be at least ${minAcceptable}`)
  }
  if (amount > manager.remainingTokens) {
    throw new Error('Insufficient tokens for this bid')
  }
  if (soldCountForManager + groupSize > manager.maxPlayers) {
    throw new Error(`${manager.name} has already reached its ${manager.maxPlayers}-player limit`)
  }
}

export function computeCommonPurseUpdate(
  team: Pick<TeamManagerEntry, 'name' | 'tokensSpent'>,
  newPurse: number,
): { purse: number; remainingTokens: number } {
  if (newPurse < team.tokensSpent) {
    throw new Error(
      `Common purse (${newPurse}) is below tokens already spent (${team.tokensSpent}) by ${team.name}`,
    )
  }
  return { purse: newPurse, remainingTokens: newPurse - team.tokensSpent }
}

export function assertUnsoldAssignment(
  player: Pick<Player, 'status'>,
  team: Pick<TeamManagerEntry, 'name' | 'maxPlayers' | 'remainingTokens'>,
  amount: number,
  soldCountForManager: number,
  // Same combo-lot meaning as assertValidBid's groupSize.
  groupSize = 1,
): void {
  if (player.status !== 'unsold') {
    throw new Error('Only unsold players can be assigned directly to a team')
  }
  if (amount < 0) {
    throw new Error('Assignment price cannot be negative')
  }
  if (soldCountForManager + groupSize > team.maxPlayers) {
    throw new Error(`${team.name} has already reached its ${team.maxPlayers}-player limit`)
  }
  if (amount > team.remainingTokens) {
    throw new Error('Insufficient tokens for this assignment')
  }
}

File: android/app/src/main/java/com/mplauction/android/data/AuctionRules.kt
package com.mplauction.android.data

import com.mplauction.android.data.model.Player
import com.mplauction.android.data.model.PlayerStatus
import com.mplauction.android.data.model.TeamManagerEntry

// Port of src/lib/auctionRules.ts. Combo lots (multiple players sold as one
// group) aren't supported yet on Android — every call site here treats
// groupSize as 1, unlike the web app's version which threads a real group
// size through for combo bids/sales.
object AuctionRules {
  fun minAcceptableBid(player: Player, bidIncrement: Long): Long =
    if (player.currentBid > 0) player.currentBid + bidIncrement else player.basePrice

  class InvalidBidException(message: String) : Exception(message)

  fun assertValidBid(
    player: Player,
    manager: TeamManagerEntry,
    amount: Long,
    bidIncrement: Long,
    soldCountForManager: Int,
  ) {
    if (player.status != PlayerStatus.open && player.status != PlayerStatus.active) {
      throw InvalidBidException("Player is not open for bidding")
    }
    val minAcceptable = minAcceptableBid(player, bidIncrement)
    if (amount < minAcceptable) throw InvalidBidException("Bid must be at least $minAcceptable")
    if (amount > manager.remainingTokens) throw InvalidBidException("Insufficient tokens for this bid")
    if (soldCountForManager + 1 > manager.maxPlayers) {
      throw InvalidBidException("${manager.name} has already reached its ${manager.maxPlayers}-player limit")
    }
  }
}
